File: diff_tool.py
from pathlib import Path
import difflib
from check_file_permission import check_file_permission

def diff_files(file1_path, file2_path, config):
    if not all(check_file_permission(f, "read", config) for f in [file1_path, file2_path]):
        raise PermissionError("Read access denied for one or more files")
    file1_path, file2_path = Path(file1_path).resolve(), Path(file2_path).resolve()
    if not all(f.is_file() for f in [file1_path, file2_path]):
        raise ValueError("One or more files not found")
    with open(file1_path, "r", encoding="utf-8") as f1, open(file2_path, "r", encoding="utf-8") as f2:
        diff = difflib.unified_diff(f1.readlines(), f2.readlines(), str(file1_path), str(file2_path))
    return "".join(diff)

File: ts-tools/diff_tool.ts
import * as fs from 'fs';
import * as path from 'path';
import * as diff from 'diff';
import { checkFilePermission } from './check_file_permission';

export function diffFiles(
  file1Path: string,
  file2Path: string,
  config: any,
): Promise<string> {
  return new Promise((resolve, reject) => {
    Promise.all([
      checkFilePermission(file1Path, 'read', config),
      checkFilePermission(file2Path, 'read', config),
    ])
      .then(() => {
        const resolved1 = path.resolve(file1Path);
        const resolved2 = path.resolve(file2Path);
        Promise.all([
          fs.promises.readFile(resolved1, 'utf-8'),
          fs.promises.readFile(resolved2, 'utf-8'),
        ])
          .then(([data1, data2]) => {
            const diffResult = diff.createPatch(file1Path, data1, data2);
            resolve(diffResult);
          })
          .catch(reject);
      })
      .catch(reject);
  });
}
